### PR TITLE
Templates

### PR DESCRIPTION
**Pull request by adityacp**
question_paper

### CODE REVIEW

File: yaksh/python_code_evaluator.py [hunk 1]
```diff
@@ -4,7 +4,7 @@
 import os
 from os.path import join
 import importlib
-
+from settings import SERVER_TIMEOUT
```
Comment (prathamesh920) [inline yaksh/python_code_evaluator.py]: Why?

File: yaksh/templates/yaksh/question.html [hunk 2]
```diff
@@ -183,10 +184,9 @@ <h3 class="brand"><strong>Online Test</h3></strong>
...
-        
-         <textarea tabindex=1 rows="10" style="padding:0;height:auto; box-shadow: none;width:750px;margin-bottom:10px;overflow:hidden;border:none;" name="answer" id="answer" wrap="off" onkeydown="return catchTab(this,event)">{% if last_attempt %}{{last_attempt}}{% else %}{% if question.type == "bash" %}{% else %}{% endif %}{% endif %}</textarea>
+        <h3>Program:</h3>
+        <textarea rows="1" class="bash" readonly="yes"  name="snippet" id="snippet" wrap="off" >{% if last_attempt %}{{ question.snippet }}{% else %}{% if question.type == "bash" %} #!/bin/bash&#13;&#10;{{ question.snippet }}{% else %}{{ question.snippet }}{% endif %}{% endif %}</textarea>
+        <textarea rows="10" class="bash" name="answer" id="answer" wrap="off" onkeydown="return catchTab(this,event)">{% if last_attempt %}{{last_attempt}}{% else %}{% if question.type == "bash" %}{% else %}{% endif %}{% endif %}</textarea>
```
Comment (prathamesh920) [inline yaksh/templates/yaksh/question.html]: If you remove  the following 
{% if question.type == "bash" %}{% else %}{% endif %}
is that fine?

**Comment by prathamesh920** [comment]
Looks fine

File: yaksh/python_code_evaluator.py (hunk 1 above)
Comment (prabhuramachandran) [inline yaksh/python_code_evaluator.py]: Aditya, please remove this import as it is not needed.

File: yaksh/python_code_evaluator.py [hunk 3]
```diff
@@ -29,6 +29,9 @@ def _check_code(self):
             fname, lineno, func, text = info[-1]
             text = str(test_code).splitlines()[lineno-1]
             err = "{0} {1} in: {2}".format(type.__name__, str(value), text)
+        except (SyntaxError, TypeError, NameError, IndentationError, ValueError) :
```
Comment (prabhuramachandran) [inline yaksh/python_code_evaluator.py]: NO, do not do this.  Instead catch the TimeoutException first and raise again, else catch the base Exception class. I had mentioned this earlier.

**Comment by prabhuramachandran** [comment]
Aditya, please add commits addressing all issues raised and we can merge thereafter.

File: yaksh/evaluator_tests/test_python_evaluation.py [hunk 4]
```diff
@@ -7,47 +7,105 @@ class PythonEvaluationTestCases(unittest.TestCase):
...
-                                 "test_id": u'null', 
-                                 "pos_args": ["3", "2"], 
-                                 "kw_args": {}
+        user_answer = "def add(a, b):\n\twhile True:\n\t\tpass"
+        timeout_msg = ("Code took more than {0} seconds to run. "
+                       "You probably have an infinite loop in your code.").format(SERVER_TIMEOUT)
+        evaluator = PythonCodeEvaluator(self.test_case_data, self.test, self.language, user_answer, ref_code_path=None, in_dir=None)
+        result = evaluator.evaluate()
+        self.assertFalse(result.get("success"))
+        self.assertEquals(result.get("error"), timeout_msg)
```
Comment (prabhuramachandran) [inline yaksh/evaluator_tests/test_python_evaluation.py]: Doing string comparisons like this is going to be fragile, what if the string is changed at some point?

File: yaksh/evaluator_tests/test_python_evaluation.py [hunk 5]
```diff
@@ -7,47 +7,105 @@ class PythonEvaluationTestCases(unittest.TestCase):
...
+
+    def test_syntax_error(self):
+        user_answer = "def add(a, b);\n\treturn a+b"
+        syntax_error_msg = ('Traceback (most recent call last):\n  File '
```
Comment (prabhuramachandran) [inline yaksh/evaluator_tests/test_python_evaluation.py]: Again, I already told you in person that this seems fragile.  Isn't there a better way to do this?

**Comment by prabhuramachandran** [comment]
You seem to have pushed other changes to this PR.  If you are going to do that, at least change the title and description.

File: yaksh/evaluator_tests/test_python_evaluation.py [hunk 6]
```diff
@@ -7,47 +7,105 @@ class PythonEvaluationTestCases(unittest.TestCase):
...
+                          "name 'add' is not defined\n")
+        evaluator = PythonCodeEvaluator(self.test_case_data, self.test, self.language, user_answer, ref_code_path=None, in_dir=None)
+        result = evaluator.evaluate()
+        self.assertFalse(result.get("success"))
+        self.assertEquals(result.get("error"), name_error_msg)
+
+    def test_recursion_error(self):
+        user_answer = ("def fact(a):\n\tif a == 0:\n\t\treturn fact(1)"
+                       "\n\telse:\n\t\treturn a * fact(a-1)")
+        self.test_case_data = [{"func_name": "fact",
```
Comment (prabhuramachandran) [inline yaksh/evaluator_tests/test_python_evaluation.py]: Why is this changed?  Why is self.test_case_data being touched at all?

File: yaksh/evaluator_tests/test_python_evaluation.py [hunk 7]
```diff
@@ -7,47 +7,105 @@ class PythonEvaluationTestCases(unittest.TestCase):
...
+        user_answer = ""
+        name_error_msg = ("Traceback (most recent call last):\nNameError: "
+                          "name 'add' is not defined\n")
+        evaluator = PythonCodeEvaluator(self.test_case_data, self.test, self.language, user_answer, ref_code_path=None, in_dir=None)
+        result = evaluator.evaluate()
+        self.assertFalse(result.get("success"))
+        self.assertEquals(result.get("error"), name_error_msg)
+
+    def test_recursion_error(self):
+        user_answer = ("def fact(a):\n\tif a == 0:\n\t\treturn fact(1)"
```
Comment (prabhuramachandran) [inline yaksh/evaluator_tests/test_python_evaluation.py]: Use `textwrap.dedent` with a triple quoted string as it is much easier to read.
